### PR TITLE
wgsl: Implement `transpose` validation tests

### PR DESCRIPTION
**Pull request by zoddicus**
<hr>

**Requirements for PR author:**

- [ ] All missing test coverage is tracked with "TODO" or `.unimplemented()`.
- [ ] New helpers are `/** documented */` and new helper files are found in `helper_index.txt`.
- [ ] Test behaves as expected in a WebGPU implementation. (If not passing, explain above.)
- [ ] Test have be tested with compatibility mode validation enabled and behave as expected. (If not passing, explain above.)

**Requirements for [reviewer sign-off](https://github.com/gpuweb/cts/blob/main/docs/reviews.md):**

- [ ] Tests are properly located in the test tree.
- [ ] [Test descriptions](https://github.com/gpuweb/cts/blob/main/docs/intro/plans.md) allow a reader to "read only the test plans and evaluate coverage completeness", and accurately reflect the test code.
- [ ] Tests provide complete coverage (including validation control cases). **Missing coverage MUST be covered by TODOs.**
- [ ] Helpers and types promote readability and maintainability.

When landing this PR, be sure to make any necessary issue status updates.

### CODE REVIEW

File: src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+} from './const_override_validation.js';
+
+export const g = makeTestGroup(ShaderValidationTest);
+
+const kValidArgumentTypes = objectsToRecord(kAllMatrices);
+
+g.test('values')
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() error on invalid inputs.
```
Comment (toji) [inline src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts:25]: Given that we don't pass any invalid inputs here, perhaps the description should be updated? (David has been catching me on the inverse so many times.)

File: src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+g.test('args')
+  .desc(`Test compilation failure of ${builtin} with variously shaped and typed arguments`)
+  .params(u => u.combine('arg', keysOf(kArgCases)))
+  .fn(t => {
+    t.expectCompileResult(
+      t.params.arg === 'good',
+      `const c = ${builtin}${kArgCases[t.params.arg]};`
+    );
+  });
+
```
Comment (toji) [inline src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts:83]: On https://github.com/gpuweb/cts/pull/3626 David pointed me at https://github.com/gpuweb/cts/pull/3612/files#diff-fe62cb0fc68ba76c7b93bf333a9a9542cdb242f14292bab49db741f6b4f72409R46 as an example of another type of validation test that we should be adding for the return types.

Seems especially relevant here, given the relationship between the input type and return type.

**Review by toji — COMMENTED** [review]
LGTM, though I think we probably want one additional test here.

File: src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts (hunk 2 above)
Comment (dneto0) [reply src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts:83]: I agree.
I feel this test should iterate over all matrix types for inputs an for outputs, and the transpose return type should be checked.

```
for input-scalar in abstract-float, f32, f16:
   for input-C in 2,3,4 
     for input-R in 2,3,4
        for output-scalar in  f32, f16:
           for output-c in 2,3,4
               for output-r in 2,3,4
                 check   something like `const c : output-mat-type = transpose(input-mat-type)` compiles or doesn't compile based on convertibility.  input-scalar-to-output-scalar, and output-r = input-C and output-c = input-r.
```

Note you can use isConvertible from src/webpgu/util/conversion.ts   (I'd probably only check convertibility of the scalar types.)

**Review by dneto0 — CHANGES_REQUESTED** [review]
Mainly elaborated on what @toji said

File: src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts (hunk 2 above)
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts:90]: Done

File: src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts (hunk 1 above)
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts]: Done

**Comment by zoddicus** [comment]
PTAL

**Review by toji — APPROVED** [review]
LGTM, thanks!

File: src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+  .desc(
+    `
+Validates that constant evaluation and override evaluation of ${builtin}() accept valid inputs.
+`
+  )
+  .params(u =>
+    u
+      .combine('stage', kConstantAndOverrideStages)
+      .combine('type', keysOf(kValidArgumentTypes))
+      .filter(u => stageSupportsType(u.stage, kValidArgumentTypes[u.type]))
```
Comment (dneto0) [inline src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts:39]: All stages support matrix math. So I think you can skip this filter.

**Review by dneto0 — APPROVED** [review]
LGTM with possible nit?

File: src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts (hunk 3 above)
Comment (zoddicus) [reply src/webgpu/shader/validation/expression/call/builtin/transpose.spec.ts]: Done

**Pull request merged**